### DISCUSSION TITLE
What is a key?

### DISCUSSION THREAD

**Discussion by faassen** [unicode-org/icu4x#3956] [Q&A] [answered]
`icu4x-datagen` offers a `--keys` option. I'm interested in generating just collation information, but I need to cover a wide range of languages as I need to build a general collation lookup mechanism. My hope is that I can slice down the generated data by controlling the keys that get generated by using specific keys like `collator/data@1`. But I have in fact no idea what keys are required for what; is there a way for me to find out? I've done some digging but haven't found enlightenment yet.

(The alternative is that I generate Rust code and rely on dead code elimination. Would that work if I do care about all possible locales?)

**Comment by Manishearth** [comment, marked as answer]
Generally we just recommend using `--keys-for-bin` to have it detect the keys you need from the binary. If this is tricky in your situation we'd love to know!

I do have an open issue about being clearer about this https://github.com/unicode-org/icu4x/issues/2685


Currently if you want to figure out which keys are needed, check out the `_unstable` constructors of each type you are constructing, look at the marker bounds, and see what the value of [KEY](https://unicode-org.github.io/icu4x/docs/icu_provider/marker/trait.KeyedDataMarker.html#associatedconstant.KEY) is.

**Reply by faassen** [reply]
Thanks!

I'm building a library that implements a programming language that needs access dynamic to collations (language-driven). I figured `--keys-for-bin` was not going to work for that use case, but I'd love to be mistaken.

**Reply by Manishearth** [reply]
It will work: language-driven is fine, "keys" are for the _type_ of data, not the language/locale stuff. You can still select the languages you need with the other flags on the datagen command line.

ICU4X is structured such that for a normal binary keys-for-bin will always contain all the keys that the program may need.

**Comment by sffc** [comment]
To add to @Manishearth's answer, `--keys-for-bin` is best if you get it to work, but to do it manually, first look at the docs:

![image](https://github.com/unicode-org/icu4x/assets/1145762/00bd420d-c1d1-45b0-9fa0-4bf481af0668)

then click on each marker bound, like [CollationSpecialPrimariesV1Marker](https://unicode-org.github.io/icu4x/docs/icu_collator/provider/struct.CollationSpecialPrimariesV1Marker.html#), and in the documentation it tells you the key string, like “collator/prim@1”
